oap-server/server-tools/data-generator/src/main/java/org/apache/skywalking/restapi/SegmentGeneratorHandler.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.skywalking.oap.server.library.module.ModuleManager;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.concurrent.Future;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class SegmentGeneratorHandler {
+    private final SourceReceiver sourceReceiver;
+    private final Map<String, Future<?>> futures = new ConcurrentHashMap<>();
+    private final EventLoopGroup eventLoopGroup = EventLoopGroups.newEventLoopGroup(10);
+
+    public SegmentGeneratorHandler(ModuleManager manager) {
+        sourceReceiver = manager.find(CoreModule.NAME).provider().getService(SourceReceiver.class);
+    }
+
+    @Post("/mock-data/segments/tasks")
+    public HttpResponse generateMockSegments(
+        @Default("0") @Param("size") int size,
+        @Default("0") @Param("qps") int qps,
+        @RequestObject SegmentRequest request) {
+
+        if (size > 0 && qps > 0) {
+            return HttpResponse.of(
+                HttpStatus.BAD_REQUEST,
+                MediaType.PLAIN_TEXT,
+                "size and qps can't be both set");
+        }
+        log.info("Generate {} mock segments, qps: {}, template: {}", size, qps, request);
+
+        final IntConsumer generator = unused -> {
+            final Segment segment = request.next();
+            log.debug("Generating segment: {}", segment);
+            sourceReceiver.receive(segment);
+        };
+        final String requestId = UUID.randomUUID().toString();
+        final Future<?> future;
+        if (size > 0) {
+            future = eventLoopGroup.submit(() -> IntStream
+                .range(0, size)
+                .forEach(generator));
+        } else {
+            future = eventLoopGroup.scheduleAtFixedRate(() -> IntStream
+                .range(0, qps)
+                .forEach(generator), 0, 1, TimeUnit.SECONDS);
+        }
+
+        futures.put(requestId, future);
+
+        future.addListener(f -> {
+            if (f.isDone()) {
+                futures.remove(requestId);
+                log.info("Generate mock segments finished: {}, requestId: {}", f.isSuccess(),
+                    requestId);
+            }
+            if (f.cause() != null) {
+                log.error(f.cause().getMessage(), f.cause());
+            }
+        });
+
+        return HttpResponse.of(MediaType.PLAIN_TEXT, requestId);
+    }
+
+    @Delete("/mock-data/segments/tasks")
+    public HttpResponse cancelRequest(@Param("requestId") String requestId) {
+        final Future<?> future = futures.get(requestId);
+        if (future == null) {
+            return HttpResponse.of(
+                HttpStatus.NOT_FOUND,
+                MediaType.PLAIN_TEXT_UTF_8,
+                "No such request: %s", requestId);
+        }
+        log.info("Cancelling request: {}", requestId);
+        future.cancel(true);
+        return HttpResponse.of(HttpStatus.OK);
+    }
+
+    @ProducesJson
+    @Get("/mock-data/segments/tasks")
```

Comment:
We should add an endpoint to expose the progress of a request. It's about how many tasks are completed, how many errors, etc.

Reply:
> We should add an endpoint to expose the progress of a request. It's about how many tasks are completed, how many errors, etc.

If you specify the `size`, there might be a progress, but if you specify `qps`, there is no progress, it generates the data endlessly until you cancel the task

Reply:
endless could have a progress for now, I think?

Reply:
Progress is just a statistic of generated data, not has to be a completion rate.

Reply:
`qps` model could show a success/error ratio instead of a simple progress bar.

Reply:
The response could be 

```json
{
"total":"int"
"running":"int"
"errors": "int"
}
```

Reply:
> The response could be
> 
> ```json
> {
> "total":"int"
> "running":"int"
> "errors": "int"
> }
> ```

There is no `running` and `errors`, data are generated and directly fed to the receiver (asynchronously), there is no response or callback from the receiver so there is no status (success / failure) of the data, all we have is the total number of generated data

Reply:
all right. `running` and `error` are out. Let's focus on the `total` now